[ews-build.webkit.org] Support PRs with many files changed

JonWBedard opened the pull request:
#### 71af31493956aee8c303ef7443e8d3a74486730b
<pre>
[ews-build.webkit.org] Support PRs with many files changed
<a href="https://bugs.webkit.org/show_bug.cgi?id=250714">https://bugs.webkit.org/show_bug.cgi?id=250714</a>
rdar://104339227

Reviewed by Aakash Jain.

GitHub&apos;s PR file changed API only provides a list of 30 files by default.
If we support pagination, we can get many more.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits._get_pr_files): Paginate files changed.

Canonical link: <a href="https://commits.webkit.org/259213@main">https://commits.webkit.org/259213@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/690d23c7d0b471c5685919e3e7528db2fe201dce

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/104375 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/77/builds/13456 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/37286 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/113592 "Built successfully") | [  ~~🛠 🧪 win~~](https://ews-build.webkit.org/#/builders/10/builds/173885 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") 
| [❌ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/108299 "Failed to checkout and rebase branch from PR 8724") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/76/builds/14554 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/85/builds/4371 "Built successfully") | [❌ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/96599 "Failed to checkout and rebase branch from PR 8724") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/112632 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/110142 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/78/builds/11217 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/3/builds/94285 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/36/builds/96599 "Failed to checkout and rebase branch from PR 8724") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/93080 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/73/builds/25896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/36/builds/96599 "Failed to checkout and rebase branch from PR 8724") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/81/builds/6816 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/70/builds/27253 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/82/builds/6946 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/84/builds/3799 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🧪 services](https://ews-build.webkit.org/#/builders/20/builds/103219 "Passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/80/builds/12971 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/62/builds/46811 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/79/builds/8739 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/3355 "Built successfully and passed tests") | | | | 
<!--EWS-Status-Bubble-End-->

ryanhaddad commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,32 @@ def _get_commit_msg(self, repo, sha):
-
-        log.msg('Failed fetching PR files: response code {}'.format(res.code))
-        return []
+        page = 1
```
Is this variable needed? I don't see it being used anywhere.

JonWBedard replied:
That means this PR isn't actually working, then....a moment to update and try and exercise the change locally

JonWBedard replied:
Updated. We do need the `page` variable because if more than 100 files are changed, GitHub will force pagination.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
 
     @defer.inlineCallbacks
     def _get_pr_files(self, repo, number):
-        # Copied from https://github.com/buildbot/buildbot/blob/v2.10.5/master/buildbot/www/hooks/github.py to include added/modified/deleted
```
Probably a good idea to keep this url, maybe change "Copied" to "Referenced"

aj062 replied:
This comment seems unresolved

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
+                ), headers=headers,
+                logger=log.msg,
+            )
+            if not response or response.status_code // 100 != 2:
```
should log something here for debugging

JonWBedard replied:
We do, a `response` of `None` will end up with a log message from `TwistedAdditions.request` (usually means the request itself failed, and `TwistedAdditions.request` knows why that might have happened), and then we get an unconditional log outside of the loop when we have an empty list, which will include the response code (if there is one)

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
-        return []
+        page = 1
+        files = []
+        while page < 30:
```
why 30? better to store this 30 in a variable describing the purpose.

JonWBedard replied:
Because GitHub advertises this API will stop working past 3000 files changed.

aj062 replied:
ok, good to add that info in a comment next to NUM_PAGE_LIMIT variable.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
     def _get_pr_files(self, repo, number):
-        # Copied from https://github.com/buildbot/buildbot/blob/v2.10.5/master/buildbot/www/hooks/github.py to include added/modified/deleted
-        headers = {"User-Agent": "Buildbot"}
+        PER_PAGE = 100
```
Nit: `PER_PAGE` => `PER_PAGE_LIMIT` (or something descriptive)

Similar to https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L171 (GitHubMixin. PER_PAGE_LIMIT and NUM_PAGE_LIMIT)

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
     @defer.inlineCallbacks
     def _get_pr_files(self, repo, number):
-        # Copied from https://github.com/buildbot/buildbot/blob/v2.10.5/master/buildbot/www/hooks/github.py to include added/modified/deleted
-        headers = {"User-Agent": "Buildbot"}
```
we don't need User-Agent anymore?

JonWBedard replied:
Because our `TwistedAdditions` code handles that for us.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -315,26 +315,36 @@ def _get_commit_msg(self, repo, sha):
+            break
+
+        if not files:
+            log.msg('Failed fetching PR files: response code {}'.format(response.code if response else '?'))
```
Might be good to include PR number in this log statement.

aj062 commented on Tools/CISupport/ews-build/events.py:
```diff
@@ -264,26 +264,38 @@ def _get_commit_msg(self, repo, sha):
+                break
+            data = response.json()
+            files += [self.file_with_status_sign(f) for f in data]
+            if len(data) == PER_PAGE_LIMIT:
```
similar to https://github.com/WebKit/WebKit/blob/main/Tools/CISupport/ews-build/steps.py#L237, we can reverse this to something like:
```
if len(data) < PER_PAGE_LIMIT:
    break
page += 1
```

aj062 reviewed (approved):
r+ with comments.

JonWBedard committed:
[ews-build.webkit.org] Support PRs with many files changed
https://bugs.webkit.org/show_bug.cgi?id=250714
rdar://104339227

Reviewed by Aakash Jain.

GitHub's PR file changed API only provides a list of 30 files by default.
If we support pagination, we can get many more.

* Tools/CISupport/ews-build/events.py:
(GitHubEventHandlerNoEdits._get_pr_files): Paginate files changed.

Canonical link: https://commits.webkit.org/259213@main
